chains/solana/contracts/programs/ccip-offramp/src/buffering.rs

```diff
@@ -0,0 +1,275 @@
+use anchor_lang::prelude::*;
+
+use crate::CcipOfframpError;
+
+#[account]
+#[derive(InitSpace, Debug)]
+pub struct ExecutionReportBuffer {
+    #[max_len(0)]
+    data: Vec<u8>,
+    chunk_bitmap: u64,
+    total_chunks: u32,
+    chunk_length: u32,
+    report_length: u32,
+}
+
+impl ExecutionReportBuffer {
+    pub fn is_initialized(&self) -> bool {
+        self.total_chunks > 0
+    }
+
+    fn filled_chunks(&self) -> u32 {
+        self.chunk_bitmap.count_ones()
+    }
+
+    pub fn is_complete(&self) -> bool {
+        self.is_initialized() && self.filled_chunks() == self.total_chunks
+    }
+
+    pub fn bytes(&self) -> Result<&[u8]> {
+        require!(
+            self.is_complete(),
+            CcipOfframpError::ExecutionReportBufferIncomplete
+        );
+
+        Ok(&self.data)
+    }
+
+    fn initialize(&mut self, report_length: u32, chunk_length: u32) -> Result<()> {
+        require!(
+            !self.is_initialized(),
+            CcipOfframpError::ExecutionReportBufferAlreadyInitialized
+        );
+        require!(
+            chunk_length > 0 && report_length >= chunk_length,
+            CcipOfframpError::ExecutionReportBufferInvalidLength
+        );
+        self.data.resize(report_length as usize, 0);
+        self.total_chunks = (report_length + chunk_length - 1) / chunk_length;
+        require_gt!(
+            64,
+            self.total_chunks,
+            CcipOfframpError::ExecutionReportBufferChunkSizeTooSmall
+        );
+        self.chunk_length = chunk_length;
+        self.report_length = report_length;
+        Ok(())
+    }
+
+    pub fn add_chunk(&mut self, report_length: u32, chunk: &[u8], chunk_index: u8) -> Result<()> {
+        if !self.is_initialized() {
+            self.initialize(report_length, chunk.len() as u32)?;
+        }
+
+        require_eq!(
+            self.report_length,
+            report_length,
+            CcipOfframpError::ExecutionReportBufferInvalidLength
+        );
+
+        let chunk_mask = 1u64 << chunk_index;
+        require!(
+            chunk_mask & self.chunk_bitmap == 0,
+            CcipOfframpError::ExecutionReportBufferAlreadyContainsChunk
+        );
+
+        if chunk.len() as u32 > self.chunk_length {
+            // We hit the special case where the first received chunk was the last one
+            // in the buffer (terminator), which may be smaller than all others. It's okay,
+            // we can recover in place.
+            return self.recover_wrong_size(report_length, chunk, chunk_index);
+        }
+
+        require_gte!(
+            self.chunk_length,
+            chunk.len() as u32,
+            CcipOfframpError::ExecutionReportBufferInvalidChunkSize
+        );
```

Comment:
Could we actually please check that only the last chunk could be undersized, and all other chunks should have exactly the same length as the buffer's constant chunk size?